<a href="../../../index.html">Go back to index</a>

<a href="../base.html">Go back to Python Portal</a>

<head>
  <link rel="stylesheet" href="../../../cssthemes/github.css">
  <meta name="viewport" content="initial-scale=1, width=device-width">
</head>


In [2]:
import sys
print(sys.executable)
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
InteractiveShell.colors = "Linux"
InteractiveShell.separate_in = 0
from tabulate import tabulate
import pandas as pd
import matplotlib.pyplot as plte 
import os, sys


tags = pd.read_csv('tags.csv')

/Users/joshchiu/repos/csjoshc.github.io/.venv/bin/python


FileNotFoundError: [Errno 2] No such file or directory: 'tags.csv'

# Dataframe data types - strings & timestamps

## String functions
Str functions don't seem to mutate the original dataframe

# Change dir to the folder this .ipynb file is in
tags = pd.read_csv('tags.csv')
```

    /home/jcmint/anaconda3/envs/learningenv/bin/python

# Dataframe data types - strings & timestamps

## String functions

Str functions don't seem to mutate the original dataframe

### Splitting columns of dataframe

In [ ]:
tags['tag'].str.split(' ').head()

0      [Mark, Waters]
1        [dark, hero]
2        [dark, hero]
3    [noir, thriller]
4        [dark, hero]
Name: tag, dtype: object

### Testing for string contents

In [ ]:
tags['isdark'] = tags['tag'].str.contains('dark') 
tags.head()

userId  movieId            tag   timestamp isdark
0      18     4141    Mark Waters  1240597180  False
1      65      208      dark hero  1368150078   True
2      65      353      dark hero  1368150079   True
3      65      521  noir thriller  1368149983  False
4      65      592      dark hero  1368150078   True

### Replace strings with other strings

In [ ]:
tags['light tag'] = tags['tag'].str.replace('dark', 'light').head()
tags.head()

userId  movieId            tag   timestamp isdark      light tag
0      18     4141    Mark Waters  1240597180  False    Mark Waters
1      65      208      dark hero  1368150078   True     light hero
2      65      353      dark hero  1368150079   True     light hero
3      65      521  noir thriller  1368149983  False  noir thriller
4      65      592      dark hero  1368150078   True     light hero

### Match regex

**Code:** `df['col1'].str.extract('*')`
Get the string that matches the reg expression

In [ ]:
tags['first tag'] = tags['tag'].str.extract('([a-zA-Z][A-Za-z])') 
tags.head()

userId  movieId            tag   timestamp isdark      light tag first tag
0      18     4141    Mark Waters  1240597180  False    Mark Waters        Ma
1      65      208      dark hero  1368150078   True     light hero        da
2      65      353      dark hero  1368150079   True     light hero        da
3      65      521  noir thriller  1368149983  False  noir thriller        no
4      65      592      dark hero  1368150078   True     light hero        da

### Split str column

In [ ]:
output = tags['light tag'].str.split(' ', expand =True) 
output.head()

0         1
0   Mark    Waters
1  light      hero
2  light      hero
3   noir  thriller
4  light      hero

### Cleanup

In [ ]:
tags.drop(tags.columns[4:7], axis = 1, inplace=True)
tags.head()

userId  movieId            tag   timestamp
0      18     4141    Mark Waters  1240597180
1      65      208      dark hero  1368150078
2      65      353      dark hero  1368150079
3      65      521  noir thriller  1368149983
4      65      592      dark hero  1368150078

## Dealing with timestamps
Need to convert UNIX POSIX timestamp to Python format before using it. 

### Cleanup

```python
tags.drop(tags.columns[4:7], axis = 1, inplace=True)
tags.head()
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }

</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>userId</th>
      <th>movieId</th>
      <th>tag</th>
      <th>timestamp</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>18</td>
      <td>4141</td>
      <td>Mark Waters</td>
      <td>1240597180</td>
    </tr>
    <tr>
      <th>1</th>
      <td>65</td>
      <td>208</td>
      <td>dark hero</td>
      <td>1368150078</td>
    </tr>
    <tr>
      <th>2</th>
      <td>65</td>
      <td>353</td>
      <td>dark hero</td>
      <td>1368150079</td>
    </tr>
    <tr>
      <th>3</th>
      <td>65</td>
      <td>521</td>
      <td>noir thriller</td>
      <td>1368149983</td>
    </tr>
    <tr>
      <th>4</th>
      <td>65</td>
      <td>592</td>
      <td>dark hero</td>
      <td>1368150078</td>
    </tr>
  </tbody>
</table>
</div>

## Dealing with timestamps

Need to convert UNIX POSIX timestamp to Python format before using it.

### Convert timestamp

In [ ]:
tags['parsed time'] = pd.to_datetime(tags['timestamp'], unit='s') 
tags.head()

userId  movieId            tag   timestamp         parsed time
0      18     4141    Mark Waters  1240597180 2009-04-24 18:19:40
1      65      208      dark hero  1368150078 2013-05-10 01:41:18
2      65      353      dark hero  1368150079 2013-05-10 01:41:19
3      65      521  noir thriller  1368149983 2013-05-10 01:39:43
4      65      592      dark hero  1368150078 2013-05-10 01:41:18

### Using timestamp to filter

In [ ]:
tags['After Dec 31 2013'] = tags['parsed time'] > '2013-12-31' 
tags.head()

userId  movieId            tag   timestamp         parsed time  \
0      18     4141    Mark Waters  1240597180 2009-04-24 18:19:40   
1      65      208      dark hero  1368150078 2013-05-10 01:41:18   
2      65      353      dark hero  1368150079 2013-05-10 01:41:19   
3      65      521  noir thriller  1368149983 2013-05-10 01:39:43   
4      65      592      dark hero  1368150078 2013-05-10 01:41:18   

   After Dec 31 2013  
0              False  
1              False  
2              False  
3              False  
4              False

### Sort based on time 
`df.sort_values(by ='col1', ascending = False)`

In [ ]:
tags = tags.sort_values(by = 'parsed time', ascending =False) 
tags.head()

NameError: name 'tags' is not defined